File: azkar/apple/macOS/Window/TodayPage.swift
import AppKit
import SwiftUI

/// Reminders on/off, the next card, and today's progress through the morning and evening lists.
struct TodayPage: View {
  @ObservedObject var model: SettingsModel
  private var s: AppStatus { model.status }

  var body: some View {
    Form {
      Section {
        HStack(spacing: 14) {
          Image(nsImage: NSApp.applicationIconImage).resizable().frame(width: 52, height: 52)
          VStack(alignment: .leading, spacing: 3) {
            Text(s.paused ? "Reminders are paused" : "Reminders are on").font(.title3.weight(.semibold))
            Text(subtitle).foregroundStyle(.secondary).monospacedDigit()
          }
          Spacer()
          Toggle("Reminders", isOn: Binding(get: { !s.paused }, set: { model.perform?(.setPaused(!$0)) }))
            .toggleStyle(.switch)
            .labelsHidden()
            .controlSize(.large)
        }
        .padding(.vertical, 4)
      }

      Section("Today") {
        progress("Morning azkar", "sunrise.fill", .orange, s.sabahDone, s.sabahTotal, .sabah)
        progress("Evening azkar", "moon.stars.fill", .purple, s.masaaDone, s.masaaTotal, .masaa)
      }

      Section {
        HStack {
          Button {
            model.perform?(.showNow)
          } label: {
            Label("Show a zikr now", systemImage: "sparkles")
          }
          Button {
            model.perform?(.dismissAll)
          } label: {
            Label("Dismiss all", systemImage: "xmark.circle")
          }
          .disabled(s.onScreen == 0)
          Spacer()
          Text(s.onScreen == 1 ? "1 card on screen" : "\(s.onScreen) cards on screen")
            .foregroundStyle(.secondary)
        }
      }

      if !s.problems.isEmpty {
        Section("Needs attention") {
          ForEach(s.problems, id: \.self) { problem in
            Label(problem, systemImage: "exclamationmark.triangle.fill").foregroundStyle(.orange)
          }
        }
      }
    }
    .formStyle(.grouped)
    .navigationTitle("Today")
  }

  private var subtitle: String {
    if s.paused { return "No new cards until you switch them back on." }
    guard let n = s.nextIn else { return "" }
    let next = String(format: "Next zikr in %d:%02d", n / 60, n % 60)
    return s.lastSkip.map { "\(next)  ·  last one skipped: \($0)" } ?? next
  }

  private func progress(
    _ title: String, _ symbol: String, _ color: Color, _ done: Int, _ total: Int,
    _ session: Session
  ) -> some View {
    HStack(spacing: 10) {
      Tile(symbol: symbol, color: color)
      VStack(alignment: .leading, spacing: 6) {
        HStack {
          Text(title)
          Spacer()
          Text(total > 0 && done >= total ? "Done" : "\(done) of \(total)")
            .foregroundStyle(.secondary)
            .monospacedDigit()
        }
        ProgressView(value: Double(min(done, total)), total: Double(max(total, 1)))
          .tint(color)
      }
      Button("Restart") { model.perform?(.restart(session)) }
        .controlSize(.small)
        .disabled(done == 0)
        .help("Start today's list again from the first zikr")
    }
    .padding(.vertical, 2)
  }
}

File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/ui/TodayPage.kt
// Reminders on or off, the next card, and today's progress through the morning and evening lists.
package com.wildduck.azkar.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.wildduck.azkar.app.Features
import com.wildduck.azkar.app.UiState
import com.wildduck.azkar.core.Session

@Composable
fun TodayPage(ui: UiState, features: Features, actions: AzkarActions) {
    Page("Today") {
        Section {
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Column(Modifier.weight(1f)) {
                    Text(
                        if (ui.paused) "Reminders are paused" else "Reminders are on",
                        style = MaterialTheme.typography.titleMedium,
                    )
                    Text(
                        subtitle(ui),
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        style = MaterialTheme.typography.bodyMedium,
                    )
                }
                Switch(!ui.paused, { actions.setPaused(!it) }, Modifier.testTag("switch:reminders"))
            }
        }

        Section("Today") {
            ProgressRow("Morning azkar", Accents.orange, ui.sabahDone, ui.sabahTotal) {
                actions.restart(Session.Sabah)
            }
            RowDivider()
            ProgressRow("Evening azkar", Accents.purple, ui.masaaDone, ui.masaaTotal) {
                actions.restart(Session.Masaa)
            }
        }

        Section {
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                Button(onClick = { actions.showNow() }) { Text("Show a zikr now") }
                if (features.cards) {
                    OutlinedButton(onClick = { actions.dismissAll() }, enabled = ui.onScreen > 0) {
                        Text("Dismiss all")
                    }
                    Spacer(Modifier.weight(1f))
                    Text(
                        if (ui.onScreen == 1) "1 card on screen" else "${ui.onScreen} cards on screen",
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        }

        if (ui.problems.isNotEmpty()) {
            Section("Needs attention") {
                for (problem in ui.problems) {
                    SettingsRow(problem.text, MaterialTheme.colorScheme.error) {
                        problem.action?.let { action ->
                            TextButton(onClick = { actions.perform(action) }) { Text(action.label) }
                        }
                    }
                }
            }
        }
    }
}

private fun subtitle(ui: UiState): String {
    if (ui.paused) return "No new cards until you switch them back on."
    val seconds = ui.nextIn ?: return ""
    val next = "Next zikr in ${seconds / 60}:${(seconds % 60).toString().padStart(2, '0')}"
    return ui.lastSkip?.let { "$next  ·  last one skipped: $it" } ?: next
}

@Composable
private fun ProgressRow(title: String, color: Color, done: Int, total: Int, onRestart: () -> Unit) {
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Tile(color)
        Column(Modifier.weight(1f)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(title, Modifier.weight(1f), style = MaterialTheme.typography.bodyLarge)
                Text(
                    if (total > 0 && done >= total) "Done" else "$done of $total",
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    style = MaterialTheme.typography.bodyMedium,
                )
            }
            LinearProgressIndicator(
                progress = { if (total > 0) (done.coerceAtMost(total).toFloat() / total) else 0f },
                modifier = Modifier.fillMaxWidth().padding(top = 6.dp),
                color = color,
                trackColor = color.copy(alpha = 0.2f),
                drawStopIndicator = {},
            )
        }
        TextButton(
            onClick = onRestart,
            modifier = Modifier.testTag("restart:$title"),
            enabled = done > 0,
        ) { Text("Restart") }
    }
}
